src/common/procedure/src/local/runner.rs

```diff
@@ -164,18 +166,49 @@ impl Runner {
             provider: self.manager_ctx.clone(),
         };
 
+        self.rolling_back = false;
+        self.execute_once_with_retry(&ctx).await;
+    }
+
+    async fn execute_once_with_retry(&mut self, ctx: &Context) {
+        let mut retry = self.exponential_builder.build();
+        let mut retry_times = 0;
         loop {
-            match self.execute_once(&ctx).await {
-                ExecResult::Continue => (),
+            match self.execute_once(ctx).await {
                 ExecResult::Done | ExecResult::Failed => return,
+                ExecResult::Continue => (),
                 ExecResult::RetryLater => {
-                    self.wait_on_err().await;
+                    retry_times += 1;
+                    if let Some(d) = retry.next() {
+                        self.wait_on_err(d, retry_times).await;
+                    } else {
+                        assert!(self.meta.state().is_retrying());
+                        if let Retrying { error } = self.meta.state() {
+                            self.meta.set_state(ProcedureState::failed(Arc::new(
+                                Error::RetryTimesExceeded {
+                                    source: error,
+                                    procedure_id: self.meta.id,
+                                },
+                            )))
+                        }
+                        return;
+                    }
                 }
             }
         }
     }
 
     async fn execute_once(&mut self, ctx: &Context) -> ExecResult {
+        // if rolling_back, there is no need to execute again.
+        if self.rolling_back {
+            if let Err(e) = self.rollback_procedure().await {
+                self.rolling_back = true;
+                self.meta.set_state(ProcedureState::retrying(Arc::new(e)));
+                return ExecResult::RetryLater;
+            }
+
+            return ExecResult::Failed;
```

Comment:
We should also set state to failed. Looks like we could wrap them into a `rollback()` method
```rust
async fn execute_once() {
    if self.rolling_back {
        return self.rollback();
    }

    // ...

    self.rollback()
}
```